exercises/practice/complex-numbers/complex-numbers_spec.lua

```diff
@@ -0,0 +1,189 @@
+local Complex = require('complex-numbers')
+
+describe('complex-numbers', function()
+  local pi = math.pi
+  local e = math.exp(1)
+  local ln = math.log
+
+  describe('real part', function()
+    it('real part of a purely real number', function()
+      assert.equal(1, Complex(1, 0).r)
+    end)
+
+    it('real part of a purely imaginary number', function()
+      assert.equal(0, Complex(0, 1).r)
+    end)
+
+    it('real part of a number with real and imaginary part', function()
+      assert.equal(1, Complex(1, 2).r)
+    end)
+  end)
+
+  describe('imaginary part', function()
+    it('imaginary part of a purely real number', function()
+      assert.equal(0, Complex(1, 0).i)
+    end)
+
+    it('imaginary part of a purely imaginary number', function()
+      assert.equal(1, Complex(0, 1).i)
+    end)
+
+    it('imaginary part of a number with real and imaginary part', function()
+      assert.equal(2, Complex(1, 2).i)
+    end)
+  end)
+
+  it('imaginary unit', function()
+    assert.equal(Complex(-1, 0), Complex(0, 1) * Complex(0, 1))
+  end)
```

Comment:
I get a failure for this:

```
Failure → ./complex-numbers_spec.lua @ 36
complex-numbers imaginary unit
./complex-numbers_spec.lua:37: Expected objects to be equal.
Passed in:
(table: 0x563a518471c0) {
  [i] = 0
  [r] = -1 }
Expected:
(table: 0x563a51846650) {
  [i] = 0
  [r] = -1 }
```

Should the spec be `assert.are.same` because it appears to be checking whether it's actually the same table, rather than whether the table's properties are the same?

Reply:
The reason it uses `assert.equal` is because it expects your implementation to define equality via the `__eq` metamethod. This lets you define what equality means for your complex type without the tests requiring an identical internal representation for two complex numbers to be considered equal. This also gives you an opportunity to compare floating point values safely (see https://www.reddit.com/r/learnprogramming/comments/9y9gxs/small_lesson_of_the_day_never_check_two_floats/).

Reply:
> The reason it uses `assert.equal` is because it expects your implementation to define equality via the `__eq` metamethod.

Got it. Thanks for the quick response!